### PR TITLE
Adhoc log improvements

### PR DESCRIPTION
**Pull request by wjordan**
This modifies the adhoc instance bootstrap process to use [CloudWatch Logs](http://docs.aws.amazon.com/AmazonCloudWatch/latest/DeveloperGuide/WhatIsCloudWatchLogs.html) API instead of the [EC2 Console Log](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-console.html#instance-console-console-output) API. This will make logging more accessible via API, and it will retain all log output (EC2 Console only retains a short buffer), and will retain them for a full day (EC2 Console only retains them for about an hour after the instance is terminated).

Also added a feature to 'tail' the CloudFormation events and CloudWatch Logs streams, so that `rake adhoc:start` and other CloudFormation-stack commands will show updated progress while waiting for a stack creation to complete.

Finally, added some `-q` flags to make the install process a bit quieter. (`chef-client` is still too verbose, but doesn't seem to be easily configurable.)

### CODE REVIEW

File: aws/cloudformation/bootstrap_standalone.sh.erb [hunk 1]
```diff
@@ -20,7 +34,24 @@ S3_BUCKET=<%=s3_bucket%>
...
+AWSLOGS_CONF=${MYTMPDIR}/awslogs.conf
+cat <<CONF > ${AWSLOGS_CONF}
+[general]
+region = ${REGION}
+[${LOG}]
+log_group_name = ${STACK}
+log_stream_name = ${LOG}
+file = ${LOG}
+CONF
+curl -s ${AWSLOGS_AGENT_URL} | python - -n -r ${REGION} -c ${AWSLOGS_CONF} > /dev/null
```
Comment (ashercodeorg) [inline aws/cloudformation/bootstrap_standalone.sh.erb:50]: Dumb Q: What is the benefit to piping to `/dev/null`?

File: lib/cdo/aws/cloud_formation.rb [hunk 2]
```diff
@@ -138,32 +143,53 @@ def update_bootstrap_script
...
 
+      # Prints the latest output from a CloudWatch Logs log stream, if present.
+      def tail_log(quiet: false)
+        log_config = {
+          log_group_name: STACK_NAME,
+          log_stream_name: LOG_NAME
+        }
+        log_config[:next_token] = @@log_token unless @@log_token.nil?
+        # Return silently if log doesn't exist.
+        resp = logs.get_log_events(log_config) rescue return
```
Comment (ashercodeorg) [inline lib/cdo/aws/cloud_formation.rb:154]: Is there a specific kind of exception we want to be rescuing?

File: lib/cdo/aws/cloud_formation.rb [hunk 3]
```diff
@@ -138,32 +143,53 @@ def update_bootstrap_script
...
         CDO.log.info "Stack #{action} requested, waiting for provisioning to complete..."
+        stack_id = STACK_NAME
         begin
-          cfn.wait_until("stack_#{action}_complete".to_sym, stack_name: STACK_NAME) do |w|
+          @@event_timestamp = start_time
+          @@log_token = nil
+          tail_log(quiet: true)
+          stack_id = cfn.describe_stacks(stack_name: stack_id).stacks.first.stack_id
+          cfn.wait_until("stack_#{action}_complete".to_sym, stack_name: stack_id) do |w|
             w.max_attempts = 360 # 1 hour
```
Comment (ashercodeorg) [inline lib/cdo/aws/cloud_formation.rb]: What units is this in? Is this meant to be `3600`? If `360` is intended, it is worth noting this.

File: aws/cloudformation/bootstrap_standalone.sh.erb (hunk 1 above)
Comment (wjordan) [inline aws/cloudformation/bootstrap_standalone.sh.erb:50]: the `awslogs-agent-setup.py` script doesn't have any `-q`/`--quiet` option and is too noisy on successful installation, so redirecting its output to `/dev/null` quiets down the output log to a more comfortable level.

**Comment by ashercodeorg** [comment]
LGTM, though I'm probably not qualified to be reviewing this code.

File: lib/cdo/aws/cloud_formation.rb (hunk 2 above)
Comment (wjordan) [inline lib/cdo/aws/cloud_formation.rb:154]: There might be a specific kind of exception for when the log doesn't exist, but `rescue return` is more concise and best describes the general intent (return silently if we can't get the log events _for any reason_). I'll update the comment to match.

**Comment by wjordan** [comment]
Thanks for taking a look! Since nobody has touched this code besides myself yet, I just wanted to make sure someone glanced over it at least.

File: lib/cdo/aws/cloud_formation.rb (hunk 3 above)
Comment (wjordan) [inline lib/cdo/aws/cloud_formation.rb]: derived from `w.delay = 10` which is seconds, so `360` max attempts is equivalent to 1 hour. i'll add a comment to clarify this

**Pull request merged**